src/utils.zig

```diff
@@ -1907,6 +1907,155 @@ test "read/write mixed" {
 	try std.testing.expect(reader.remaining.len == 0);
 }
 
+pub fn SparseSet(comptime T: type, comptime IdType: type) type { // MARK: SparseSet
+	std.debug.assert(@typeInfo(IdType) == .int);
+	std.debug.assert(@typeInfo(IdType).int.signedness == .unsigned);
+
+	return struct {
+		const Index = enum(IdType) {
+			noValue = std.math.maxInt(IdType),
+			_,
+		};
+
+		const Self = @This();
+
+		dense: main.ListUnmanaged(T) = .{},
+		reverseIds: main.ListUnmanaged(IdType) = .{},
+		sparse: main.ListUnmanaged(Index) = .{},
+
+		pub fn deinit(self: *Self, allocator: NeverFailingAllocator) void {
+			self.dense.deinit(allocator);
+			self.reverseIds.deinit(allocator);
+			self.sparse.deinit(allocator);
+		}
+
+		pub fn contains(self: *Self, id: IdType) bool {
+			return id < self.sparse.items.len and self.sparse.items[id] != .noValue;
+		}
+
+		pub fn set(self: *Self, allocator: NeverFailingAllocator, id: IdType, value: T) void {
+			if(id >= self.sparse.items.len) {
+				self.sparse.appendNTimes(allocator, .noValue, id - self.sparse.items.len + 1);
+			}
+
+			std.debug.assert(self.sparse.items[id] == .noValue);
+
+			const denseId: Index = @enumFromInt(self.dense.items.len);
```

Comment:
There is still a problem here: What if `self.dense.items.len` is `.noValue`?

Reply:
There is also no tests covering that while there should be one since this is a very important edge case.